OneSignalSDK/onesignal/src/main/java/com/onesignal/OSTaskController.java

```diff
@@ -0,0 +1,128 @@
+package com.onesignal;
+
+import androidx.annotation.NonNull;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.atomic.AtomicLong;
+
+class OSTaskController {
+
+    // the concurrent queue in which we pin pending tasks upon finishing initialization
+    static final ConcurrentLinkedQueue<Runnable> taskQueueWaitingForInit = new ConcurrentLinkedQueue<>();
+    private static final AtomicLong lastTaskId = new AtomicLong();
+    private static ExecutorService pendingTaskExecutor;
+
+    private final OSLogger logger;
+    private final OSRemoteParamController paramController;
+
+    private List<String> methodsAvailableForDelay;
+
+    OSTaskController(OSRemoteParamController paramController, OSLogger logger) {
+        this.paramController = paramController;
+        this.logger = logger;
+
+        methodsAvailableForDelay = new ArrayList<>(
+                Arrays.asList("getTags()", "SyncHashedEmail()", "setExternalUserId()", "setSubscription()",
+                        "promptLocation()", "idsAvailable()", "sendTag()", "sendTags()", "handleNotificationOpen()",
+                        "setEmail()"));
+    }
```

Comment:
good catch!

Reply:
> And if we wanna make it easier to change in the future we should create `static final String` for each method name
> 
> OR
> 
> We make a helper that calls and returns the current enclosing method name
> **Example:**
> `String name = new Object(){}.getClass().getEnclosingMethod().getName();`
> Which would return `"sendTag"` if we called this within the `sendTag` method

I like this suggestion, but Im wondering the performance cost, because that one is creating an anonymous class, and well you know that anonymous class compiled adds some extra code to the compilation, also getEnclosingMethod I think is a reflection method, and by definition reflection = not performant 😁 . I think i will continue with the constants, that seems right